leakybucket.go

```diff
@@ -391,3 +395,129 @@ func (t *LeakyBucketRedis) SetState(ctx context.Context, state LeakyBucketState)
+// PartitionKey is the key used to store all the this implementation in DynamoDB.
+//
+// TableProps describe the table that this backend should work with. This backend requires the following on the table:
+// * TTL
+//
+// TTL is the TTL of the stored item.
+//
+// If raceCheck is true and the item in DynamoDB are modified in between State() and SetState() calls then
+// ErrRaceCondition is returned.
+func NewLeakyBucketDynamoDB(client *dynamodb.Client, partitionKey string, tableProps DynamoDBTableProperties, ttl time.Duration, raceCheck bool) *LeakyBucketDynamoDB {
+	keys := map[string]types.AttributeValue{
+		tableProps.PartitionKeyName: &types.AttributeValueMemberS{Value: partitionKey},
+	}
+
+	if tableProps.SortKeyUsed {
+		keys[tableProps.SortKeyName] = &types.AttributeValueMemberS{Value: partitionKey}
+	}
+
+	return &LeakyBucketDynamoDB{
+		client:       client,
+		partitionKey: partitionKey,
+		tableProps:   tableProps,
+		ttl:          ttl,
+		raceCheck:    raceCheck,
+		keys:         keys,
+	}
+}
+
+// State gets the bucket's state from DynamoDB.
+func (t *LeakyBucketDynamoDB) State(ctx context.Context) (LeakyBucketState, error) {
+	resp, err := dynamoDBGetItem(ctx, t.client, t.getGetItemInput())
+
+	if err != nil {
+		return LeakyBucketState{}, err
+	}
+
+	return t.loadStateFromDynamoDB(resp)
+}
+
+// SetState updates the state in DynamoDB.
+func (t *LeakyBucketDynamoDB) SetState(ctx context.Context, state LeakyBucketState) error {
+	input := t.getPutItemInputFromState(state)
+
+	var err error
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		_, err = dynamoDBputItem(ctx, t.client, input)
+	}()
+
+	select {
+	case <-done:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	return err
+}
+
+const (
+	dynamodbBucketRaceConditionExpression = "Version <= :version"
+	dynamoDBBucketLastKey                 = "Last"
+	dynamoDBBucketVersionKey              = "Version"
+)
+
+func (t *LeakyBucketDynamoDB) getPutItemInputFromState(state LeakyBucketState) *dynamodb.PutItemInput {
+	item := map[string]types.AttributeValue{}
+	for k, v := range t.keys {
+		item[k] = v
+	}
+
+	item[dynamoDBBucketLastKey] = &types.AttributeValueMemberN{Value: strconv.Itoa(int(state.Last))}
+	item[dynamoDBBucketVersionKey] = &types.AttributeValueMemberN{Value: strconv.Itoa(int(t.latestVersion) + 1)}
+	item[t.tableProps.TTLFieldName] = &types.AttributeValueMemberN{Value: strconv.Itoa(int(time.Now().Add(t.ttl).Unix()))}
+
+	input := &dynamodb.PutItemInput{
+		TableName: &t.tableProps.TableName,
+		Item:      item,
+	}
+
+	if t.raceCheck && t.latestVersion > 0 {
+		input.ConditionExpression = aws.String(dynamodbBucketRaceConditionExpression)
+		input.ExpressionAttributeValues = map[string]types.AttributeValue{
+			":version": &types.AttributeValueMemberN{Value: strconv.Itoa(int(t.latestVersion))},
+		}
+	}
+
+	return input
+}
+
+func (t *LeakyBucketDynamoDB) getGetItemInput() *dynamodb.GetItemInput {
```

Comment:
Struct LeakyBucketDynamoDB has methods on both value and pointer receivers. Such usage is not recommended by the Go Documentation.

Reply:
All methods are on pointer receivers.